include/tcpdf_cmssignature.php

```diff
@@ -0,0 +1,545 @@
+<?php
+/*
+// File name   : tcpdf_cmssignature.php
+// Version     : 1.0.1
+// Begin       : 2023-05-25
+// Last Update : 2023-05-25
+// Author      : Hida - https://github.com/hidasw
+// License     : GNU GPLv3
+*/
+/**
+ * @class tcpdf_cms_signature
+ * Manage CMS Signature for TCPDF.
+ * @version 1.0.001
+ * @author M Hida
+ */
+class tcpdf_cms_signature {
+	/**
+	 * Catch error.
+	 * @public
+	 */
+	public $errorMsg;
+	
+	/**
+	 * result value of pkcs7 EncryptedDigest.
+	 * @public
+	 */
+	public string $pkcs7_EncryptedDigest;
+	
+	/**
+	 * private array parsed of pkcs7 data.
+	 * @private
+	 */
+	private array $pkcs7_dataArray;
+
+	/**
+	 * futher implementation
+	 * @public
+	 */
+	public function __construct() {
+
+	}
+
+	/**
+	 * Create common tsa query with SHA1 digest, nonce and cert req extension
+	 * @param string $binaryData raw/binary data of tsa query
+	 * @return string binary tsa query
+	 * @public
+	 */
+	public function tsa_query($binaryData) {
+	$hash = hash('sha1', $binaryData);
+	$tsReqData = tcpdf_asn1::seq(
+															tcpdf_asn1::int(1).
+															tcpdf_asn1::seq(
+																							tcpdf_asn1::seq("06052B0E03021A"."0500"). // object OBJ_sha1
+																							tcpdf_asn1::oct($hash)).
+																							tcpdf_asn1::int(hash('crc32', rand())). // tsa nonce
+																							'0101ff' // req return cert
+															);
+	return hex2bin($tsReqData);
+	}
+
+	/**
+	 * function to get response header of specific name
+	 * @param string $headerkeyName header name eg Content-Type
+	 * @param string $headerData text of header
+	 * @return string value of header name
+	 * @protected
+	 */
+	protected function getResponseHeader($headerkeyName, $headerData) {
+		$headers = explode("\n", $headerData);
+		foreach ($headers as $key => $r) {
+			// Match the header name up to ':', compare lower case
+			if (stripos($r, $headerkeyName . ':') === 0) {
+				list($headername, $headervalue) = explode(":", $r, 2);
+				return trim($headervalue);
+			}
+		}
+	}
+
+	/**
+	 * send tsa query with curl
+	 * @param string $binarytsReqData binary ts query
+	 * @param string $tsa_host='', $tsa_username='', $tsa_password=''
+	 * @return string tsa response
+	 * @public
+	 */
+	public function tsa_send($binarytsReqData, $tsa_host='', $tsa_username='', $tsa_password='') {
+		$ch = curl_init();
+		curl_setopt($ch, CURLOPT_URL, $tsa_host);
+		if(isset($tsa_username) && isset($tsa_password)) {
+			curl_setopt($ch, CURLOPT_USERPWD, $tsa_username . ":" . $tsa_password);
+		}
+		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+		curl_setopt($ch, CURLOPT_POST, 1);
+		curl_setopt($ch, CURLOPT_HEADER, 1);
+		curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/timestamp-query','User-Agent: TCPDF'));
+		curl_setopt($ch, CURLOPT_POSTFIELDS, $binarytsReqData);
+		$tsResponse = curl_exec($ch);
+		curl_close($ch);
+
+		if($tsResponse) {
+			$header_size = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
```

Comment:
This code showed me an error because this is trying to access to curl info after doing curl_close. Can you check it?

Thank you.

Reply:
yes curl_close($ch); should be moved to after curl_getinfo . maybe a php version compatibility issue, I think this has worked before (or I forgot) $tsResponse variable should save the curl_exec($ch) resource and keep it even if it has been closed.